### PR TITLE
Device renaming during ssh|vnc AY installation

### PR DESCRIPTION
**Pull request by mchf**
bnc#944349

the most important part of the patch is moving device renaming from 2nd into 1st stage

Do not merge yet, still under testing.

### CODE REVIEW

File: src/clients/save_network.rb [hunk 1]
```diff
@@ -189,14 +190,14 @@ def copy_udev_rules
         log.info("File #{udev_rules_destdir} exists")
       end
 
-      if !FileUtils.Exists(net_destfile)
+      if !Mode.update
```
Comment (kobliha) [inline src/clients/save_network.rb]: This deserves a comment, why you copy this file only during Update?

**Comment by kobliha** [comment]
RuboCop complains

File: src/clients/save_network.rb (hunk 1 above)
Comment (mchf) [inline src/clients/save_network.rb]: I do not copy it during update to avoid overwriting user changes

**Comment by mchf** [comment]
I've finished last of "basic tests" and everything seems to be fine.

File: src/clients/save_network.rb [hunk 2]
```diff
@@ -263,6 +266,63 @@ def configure_target
...
+    #
+    # Intended mainly for steps which cannot be done in AY's second stage
+    #
+    # FIXME: Currently used only for applying udev rules during network
+    # installations (ssh, vnc, ...). It was introduced as a quick fix for
+    # bnc#944349, so it is currently limited only on {ssh|vnc} installations.
+    # Once properly analyzed and tested then starting of whole network second
+    # stage can be moved here.
+    def ay_mode_configuration
+      # TODO: run only when in {ssh|vnc} mode of AY
```
Comment (jreidinger) [inline src/clients/save_network.rb]: why is there todo, when code below do it?

File: src/clients/save_network.rb [hunk 3]
```diff
@@ -263,6 +266,63 @@ def configure_target
...
+
+      LanItems.Read
+
+      udev_rules.each do |rule|
+        name_to = rule["name"]
+        attr = rule["rule"]
+        key = rule["value"].downcase
+        # currently we're interrested only on those interfaces which are already
+        # configured - such interfaces cannot be restarted during second stage
+        matching_item = LanItems.Items.find { |_, i| i["hwinfo"]["busid"].downcase == key || i["hwinfo"]["mac"].downcase == key }
```
Comment (jreidinger) [inline src/clients/save_network.rb]: hint is to use `index, matching_item = ...` this way below you do not need to use nasty `matching_item[1]`

File: src/clients/save_network.rb (hunk 2 above)
Comment (mchf) [inline src/clients/save_network.rb]: forgot to drop it ;-)

**Comment by jreidinger** [comment]
few NP, in general looks fine for me.

File: src/modules/LanItems.rb [hunk 4]
```diff
@@ -645,7 +645,7 @@ def WriteUdevItemsRules
           # - remove it completely
           # removing is less error prone when tracking name changes, so it was chosen.
           item_udev_net = RemoveKeyFromUdevRule(item_udev_net, "KERNEL")
-          SetLinkDown(dev_name)
+          SetLinkDown(dev_name) if !Mode.autoinst
```
Comment (schubi2) [inline src/modules/LanItems.rb]: I would put here also an comment, that AutoYaST would have here a problem if it used via SSH installation.

**Comment by kobliha** [comment]
Merging, we need it NOW

**Pull request merged**